source/controllers/ChatController.ts

```diff
@@ -0,0 +1,202 @@
+				logger.error(`MessageId undefined in chat message body despite message stage being ${req.body.stage}!`);
+				res.sendStatus(500);
+				return;
+			}
+			// Quick validation for the KEY stage, just in case someone attempts to send a malformed request that isn't an actual key part.
+			if (req.body.stage === 'KEY') {
+				if (req.body.content.split('|').length !== 2 || req.body.content.split('|')[0] !== user.id) {
+					res.sendStatus(400);
+					return;
+				}
+			}
+			const chatMessage = await ChatMessage.findByPk(req.body.messageId);
+			if (chatMessage === undefined || chatMessage === null) {
+				res.sendStatus(400);
+				return;
+			}
+			await MessageComponent.create({
+				id: uuidv4(),
+				messageId: chatMessage.id,
+				stage: req.body.stage,
+				content: req.body.content
+			});
+			if (chatMessage.receiverId !== null && chatMessage.senderId !== null) {
+				if (req.body.stage === 'KEY') {
+					// Notify that a KEY component is ready
+					if (user.id === chatMessage.receiverId) io.to(chatMessage.senderId).emit('messageKey', user.id, chatMessage.id);
+					else io.to(chatMessage.receiverId).emit('messageKey', user.id, chatMessage.id);
+				} else io.to(chatMessage.receiverId).emit('messageContent', user.id, chatMessage.id); // Notify that a CONTENT component is ready
+			}
+			res.sendStatus(200);
+		}
+	}
+);
+
+const message = chat.subcontroller('message', ['id']);
+
+message.read = async (req, res) => {
+	const user = res.locals.user;
+	if (user === undefined || user === null) {
+		res.sendStatus(403); // You should be signed in to access chat
+		return;
+	}
+	const stage = req.query.stage;
+	if (stage === null || typeof stage !== 'string' || !['ANNOUNCE', 'KEY', 'MESSAGE'].includes(stage)) {
+		res.sendStatus(400); // Malformed request
+		return;
+	}
+	const chatMessage = await ChatMessage.findByPk(req.params.id);
+	if (chatMessage === undefined || chatMessage === null) {
+		res.sendStatus(400); // Invalid data
+		return;
+	}
+	if (chatMessage.senderId !== user.id && chatMessage.receiverId !== user.id) {
+		res.sendStatus(403); // Trying to snoop on someone else's convos
+		return;
+	}
+	const stageMessages = await MessageComponent.findAll({
+		where: {
+			messageId: chatMessage.id,
+			stage
+		}
+	});
+	if (stage === 'ANNOUNCE' || stage === 'MESSAGE') {
+		if (stageMessages.length === 0) {
+			res.sendStatus(404); // Not found, hopefully the client will understand that it might appear later
+		} else {
+			res.send(stageMessages[0].content);
+		}
+	} else {
+		if (stageMessages.length === 0) {
+			res.sendStatus(404); // Not found, hopefully the client will understand that it might appear later
+			return;
+		}
+		const otherKey = stageMessages.find(msg => !msg.content.includes(user.id)); // Find the key sent by the other user, if any
+		if (otherKey === undefined) {
+			res.sendStatus(404); // Not found, hopefully the client will understand that it might appear later
+		} else {
+			res.send(otherKey.content);
+		}
+	}
```

Comment:
This is the part where 404 isn't exactly appropriate. As there is no fault of the user this automatically makes it not a 400 error. `204 No content` is much more fitting.